File: examples/ch04/persistence.py
"""Persistence — answers to one actor: engineering (how orders are stored).

One reason to change: the storage format. Change to a database and only this
module moves.
"""

import json
from pathlib import Path

from models import Order

ORDERS_FILE = Path("orders.json")


def save_order(order: Order, total: float) -> None:
    record = {"customer": order.customer.name, "total": total}
    existing = json.loads(ORDERS_FILE.read_text()) if ORDERS_FILE.exists() else []
    existing.append(record)
    ORDERS_FILE.write_text(json.dumps(existing, indent=2))

File: examples/ch04/ts/persistence.ts
// Persistence — answers to one actor: engineering (how orders are stored). One
// reason to change: the storage format. Move to a database and only this module
// changes.

import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { type Order } from "./models";

interface OrderRecord {
  readonly customer: string;
  readonly total: number;
}

export const saveOrder = (
  order: Order,
  total: number,
  ordersFile = "orders.json",
): void => {
  const record: OrderRecord = { customer: order.customer.name, total };
  const existing: OrderRecord[] = existsSync(ordersFile)
    ? (JSON.parse(readFileSync(ordersFile, "utf8")) as OrderRecord[])
    : [];
  existing.push(record);
  writeFileSync(ordersFile, JSON.stringify(existing, null, 2));
};
